compiler/resolution/implementForallIntents2.cpp

```diff
@@ -0,0 +1,1097 @@
+    curFn(iterFn),
+    nested(false),
+    retSym(retSymbol),
+    anchorFn(curFn),
+    anchor1(0), anchor2(0), anchor2orig(0), anchor2prev(0)
+  {
+    fiInfos.resize(forall->numIntentVars());
+  }
+
+  // constructor for a nested context
+  FIcontext(FIcontext& parentCx, CallExpr* taskCall, FnSymbol* taskFn) :
+    forall(parentCx.forall),
+    call(taskCall),
+    curFn(taskFn),
+    nested(true),
+    retSym(parentCx.retSym),
+    anchorFn(curFn),
+    anchor1(0), anchor2(0), anchor2orig(0), anchor2prev(0)
+  {
+    int numSVars = forall->numIntentVars();
+    fiInfos.resize(numSVars);
+    for (int ix = 0; ix < numSVars; ix++) {
+      fiInfos[ix].svSymbol = parentCx.fiInfos[ix].svSymbol;
+      fiInfos[ix].fiActual = parentCx.fiInfos[ix].fiFormal;
+    }
+  }
+};
+
+
+/////////// EFLOPI: Enclosing For Loop Over Parallel Iterator ///////////
+
+// 
+// Info we gather while looking for an eflopi or an enclosing forall loop
+//
+// The contents are valid only after a call to findEnclosingParallelLoop()
+// as follows:
+//
+//  * 'eflopiForall' and 'eflopiLoop' are always set up
+//   -  to the corresponding ForallStmt or ForLoop node if one is found,
+//   -  to NULL if neither is found.
+//
+//  * the other fields are set up only if eflopiLoop != NULL.
+//
+class EflopiInfo {
+public:
+  ForallStmt* eflopiForall;
+  ForLoop*    eflopiLoop;
+  CallExpr*   eflopiCall;
+  // more details
+  // todo soon: remove all except asgnToIter?
+  Symbol*    iterSym;
+  CallExpr*  asgnToIter;
+  Symbol*    iterCallTemp;
+  CallExpr*  asgnToCallTemp;
+};
+
+//
+// Is 'forLoop' a loop over a parallel iterator?
+// If so, fill in 'eInfo' with details.
+//
+static bool findCallToParallelIterator(ForLoop* forLoop, EflopiInfo& eInfo)
+{
+  Symbol* iterSym = forLoop->iteratorGet()->symbol();
+
+  // Find an assignment to 'iterSym'.
+  CallExpr* asgnToIter = NULL;
+  for (Expr* curr = forLoop->prev; curr; curr = curr->prev)
+    if (CallExpr* call = toCallExpr(curr))
+      if (call->isPrimitive(PRIM_MOVE))
+        if (SymExpr* lhs1 = toSymExpr(call->get(1)))
+          if (lhs1->symbol() == iterSym) {
+            asgnToIter = call;
+            break;
+          }
+  INT_ASSERT(asgnToIter);
+
+  // We have:
+  //   move( call_tmp call( ITERATOR args... ) )
+  //   move( _iterator call( _getIterator call_tmp ) )
+  // We need to see if args... contain a leader or standalone tag.
+  // 'asgnToIter' is the second of the above moves. Find the first one.
+  CallExpr* rhs1 = toCallExpr(asgnToIter->get(2));
+  INT_ASSERT(rhs1 && rhs1->isNamed("_getIterator"));
+  Symbol* calltemp = toSymExpr(rhs1->get(1))->symbol();
+  CallExpr* asgnToCallTemp = NULL;
+  for (Expr* curr = asgnToIter->prev; curr; curr = curr->prev)
+    if (CallExpr* call = toCallExpr(curr))
+      if (call->isPrimitive(PRIM_MOVE))
+        if (SymExpr* lhs2 = toSymExpr(call->get(1)))
+          if (lhs2->symbol() == calltemp) {
+            asgnToCallTemp = call;
+            break;
+          }
```

Comment:
To be clear, I don't like that we have to write this kind of pattern-matching-something-lowered code, but I don't have an alternative in mind. Would it help to add something-or-other to ForLoop? Or to promote the 'getIterator' idea to its own AST node or something?